// @microsoft/sp-http/lib-commonjs/msGraphClient/MSGraphClientV3.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var sp_core_library_1 = require("@microsoft/sp-core-library");
var sp_diagnostics_1 = require("@microsoft/sp-diagnostics");
var microsoft_graph_clientV3_1 = require("@microsoft/microsoft-graph-clientV3");
var AadTokenProviders_1 = tslib_1.__importDefault(require("../oauthTokenProvider/AadTokenProviders"));
var GraphHttpClientContext_1 = tslib_1.__importDefault(require("../graphHttpClient/GraphHttpClientContext"));
/**
 * MSGraphClient is used to perform REST calls against Microsoft Graph.
 *
 * @remarks The Microsoft Graph JavaScript client library is a lightweight wrapper around the
 * Microsoft Graph API. This class allows developers to start making REST calls to MSGraph without
 * needing to initialize the the MSGraph client library. If a custom configuration is desired,
 * the MSGraphClient api function needs to be provided with that custom configuration for
 * every request.
 *
 * For more information: {@link https://github.com/microsoftgraph/msgraph-sdk-javascript}
 *
 * @public
 */
var MSGraphClientV3 = /** @class */ (function () {
    /**
     * @param serviceScope - Provides services for the MSGraphClient to consume.
     *
     * @internal
     */
    function MSGraphClientV3(serviceScope) {
        var _this = this;
        serviceScope.whenFinished(function () {
            MSGraphClientV3._graphBaseUrl = serviceScope
                .consume(GraphHttpClientContext_1.default.serviceKey)
                .getGraphEndpointUrl();
            if (!MSGraphClientV3._instance) {
                MSGraphClientV3._instance = _this._createGraphClientInstance(AadTokenProviders_1.default.configurable);
            }
        });
    }
    /**
     * All calls to Microsoft Graph are chained together starting with the api function.
     *
     * @remarks Path supports the following formats:
     * * me
     * * /me
     * * https://graph.microsoft.com/v1.0/me
     * * https://graph.microsoft.com/beta/me
     * * me/events?$filter=startswith(subject, 'ship')
     *
     * The authProvider and baseUrl option should not be used, as they have already been
     * provided by the framework. See the official documentation here:
     * https://github.com/microsoftgraph/msgraph-sdk-javascript
     *
     * @param path - The path for the request to MSGraph.
     * @param config - Sets the configuration for this request.
     */
    MSGraphClientV3.prototype.api = function (path) {
        sp_core_library_1.Validate.isNonemptyString(path, 'path');
        return MSGraphClientV3._instance.api(path);
    };
    MSGraphClientV3.prototype._createGraphClientInstance = function (tokenProvider) {
        var _this = this;
        var qosMonitor = new sp_diagnostics_1._QosMonitor('MSGraphClientV3.CreateGraphClientInstance');
        var graphClient = microsoft_graph_clientV3_1.Client.init({
            authProvider: function (done) {
                _this._getOAuthToken(tokenProvider)
                    .then(function (token) {
                    done(undefined /** error object */, token);
                    qosMonitor.writeSuccess();
                })
                    .catch(function (e) {
                    // The functions will return a rejected promise if it fails to get an OAuth2Token.
                    done(e, '');
                    qosMonitor.writeExpectedFailure(e.message);
                });
            },
            baseUrl: MSGraphClientV3._graphBaseUrl
        });
        return graphClient;
    };
    MSGraphClientV3.prototype._getOAuthToken = function (tokenProvider) {
        var qosMonitor = new sp_diagnostics_1._QosMonitor('MSGraphClient.FetchGraphToken');
        return tokenProvider
            .getToken(MSGraphClientV3._graphBaseUrl)
            .then(function (response) {
            qosMonitor.writeSuccess();
            return response;
        })
            .catch(function (error) {
            qosMonitor.writeUnexpectedFailure('FetchGraphToken', error);
            throw error;
        });
    };
    return MSGraphClientV3;
}());
exports.default = MSGraphClientV3;
