// @microsoft/sp-adaptive-card-extension-base/lib-commonjs/AdaptiveCardExtensionLoader.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AdaptiveCardExtensionLoader = void 0;
var sp_loader_1 = require("@microsoft/sp-loader");
var sp_core_library_1 = require("@microsoft/sp-core-library");
var sp_diagnostics_1 = require("@microsoft/sp-diagnostics");
var sp_telemetry_1 = require("@ms/sp-telemetry");
var AdaptiveCardExtensionContext_1 = require("./AdaptiveCardExtensionContext");
var ICachedLoadParameters_1 = require("./caching/ICachedLoadParameters");
var Killswitches_1 = require("./common/Killswitches");
var INIT_FAILURE = 'AdaptiveCardExtensionInitializationFailed';
var LOAD_FAILURE = 'AdaptiveCardExtensionLoadFailed';
/**
 * BaseAdaptiveCardExtension loader.
 *
 * @internal
 */
var AdaptiveCardExtensionLoader = /** @class */ (function () {
    function AdaptiveCardExtensionLoader() {
        // noop
    }
    AdaptiveCardExtensionLoader._createQosExtraData = function (manifest, deviceContext) {
        return {
            alias: manifest.alias,
            deviceContext: deviceContext,
            isInternal: manifest.isInternal,
            manifestId: manifest.id,
            mpnId: manifest.mpnId,
            scenarioId: (0, Killswitches_1.isLogScenarioIdKSActivated)() ? undefined : sp_telemetry_1._PerformanceLogger.getScenarioId(),
            storeAppId: manifest.storeAppId,
            teamsHosted: sp_core_library_1._BrowserUtilities.isTeamsHosted()
        };
    };
    /**
     * Load an Adaptive Card Extension.
     *
     * @remarks Returning the extension instance will enable the host to directly interact with the ACE
     * rather than using the manager as an intermediary.
     *
     * @remarks We pass the host through the context to enable the ACE to invoke functionality from the host.
     *
     * @param loadContext - See {@link _IAdaptiveCardExtensionLoadContext}
     */
    AdaptiveCardExtensionLoader.prototype.load = function (loadContext) {
        var _this = this;
        var monitor = !(0, Killswitches_1.isCopyEngagementToInteranaKSActivated)()
            ? new sp_diagnostics_1._QosMonitor('AdaptiveCardExtension.Load', true, true)
            : new sp_diagnostics_1._QosMonitor('AdaptiveCardExtension.Load', true);
        var extraData = AdaptiveCardExtensionLoader._createQosExtraData(loadContext.manifest, loadContext.deviceContext);
        // SPARC team owns an expensive operation to add new dimensions. We use 'WebPart' instead of 'AdaptiveCardExtension' to work around creating a new dimension
        var tagName = "WebPart.ACE_".concat(loadContext.manifest.alias, ".").concat(loadContext.manifest.isInternal ? 'internal' : 'external', ".").concat(loadContext.instanceId);
        var thrownError;
        var loadedAce;
        if (!(0, Killswitches_1.isCaptureACELifecyclePerformanceKSActivated)()) {
            sp_telemetry_1._PerformanceLogger.startMarkForComponent(tagName);
        }
        return sp_loader_1.SPComponentLoader.loadComponent(loadContext.manifest)
            .then(function (module) {
            // tslint:disable-next-line:no-any
            if (module.default) {
                module = module.default; // tslint:disable-line:no-any
            }
            if (!(0, Killswitches_1.isCaptureACELifecyclePerformanceKSActivated)()) {
                sp_telemetry_1._PerformanceLogger.markComponentModuleLoaded(tagName);
            }
            return new module();
        })
            .catch(function (error) {
            if (!(0, Killswitches_1.isCaptureACELifecyclePerformanceKSActivated)()) {
                sp_telemetry_1._PerformanceLogger.endFailureForComponent(tagName, 'Load', error);
            }
            monitor.writeUnexpectedFailure(LOAD_FAILURE, error, extraData);
            thrownError = error;
            throw error;
        })
            .then(function (ace) {
            loadedAce = ace;
            loadedAce._internalInitialize(_this._getContext(loadContext), loadContext.addedFromPersistedData, loadContext.initialDisplayMode, tagName);
            loadedAce._internalDeserialize(loadContext.aceData);
            loadContext.host.onBeforeInitialized(loadedAce);
            var cachedLoadParameters = loadContext.cacheObject
                ? (0, ICachedLoadParameters_1._createCachedLoadParameters)(loadContext.cacheObject)
                : undefined;
            return loadedAce.onInit(cachedLoadParameters);
        })
            .then(function () {
            if (!(0, Killswitches_1.isCaptureACELifecyclePerformanceKSActivated)()) {
                sp_telemetry_1._PerformanceLogger.markComponent(tagName, 'init');
            }
            monitor.writeSuccess(extraData);
            loadedAce._internalOnAfterInit();
            return loadedAce;
        })
            .catch(function (error) {
            if (!thrownError) {
                if (!(0, Killswitches_1.isCaptureACELifecyclePerformanceKSActivated)()) {
                    sp_telemetry_1._PerformanceLogger.endFailureForComponent(tagName, 'Initialize', error);
                }
                monitor.writeUnexpectedFailure(INIT_FAILURE, error, extraData);
            }
            throw error;
        });
    };
    AdaptiveCardExtensionLoader.prototype._getContext = function (loadContext) {
        return new AdaptiveCardExtensionContext_1.AdaptiveCardExtensionContext({
            host: loadContext.host,
            manifest: loadContext.manifest,
            instanceId: loadContext.instanceId,
            parentServiceScope: loadContext.serviceScope,
            loggingTag: '',
            deviceContext: loadContext.deviceContext,
            location: loadContext.location
        });
    };
    return AdaptiveCardExtensionLoader;
}());
exports.AdaptiveCardExtensionLoader = AdaptiveCardExtensionLoader;
