// @microsoft/spfx-heft-plugins/lib/spfxHeftPlugin.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const DevDeployPlugin_1 = __importDefault(require("./plugins/devDeployPlugin/DevDeployPlugin"));
const WebpackConfigurationPlugin_1 = __importDefault(require("./plugins/webpackConfigurationPlugin/WebpackConfigurationPlugin"));
const CustomizeWebpackConfigurationPlugin_1 = __importDefault(require("./plugins/CustomizeWebpackConfigurationPlugin"));
const WebpackServeConfigurationPlugin_1 = __importDefault(require("./plugins/webpackConfigurationPlugin/WebpackServeConfigurationPlugin"));
/**
 * @public
 */
class SpfxHeftPlugin {
    constructor() {
        this.pluginName = "spfx-heft-plugin" /* SpfxHeftPlugin */;
    }
    get accessor() {
        if (!this._accessor) {
            this._accessor = {
                devDeploy: DevDeployPlugin_1.default.accessor,
                spfxWebpackConfiguration: WebpackConfigurationPlugin_1.default.accessor,
                spfxWebpackServeConfiguration: WebpackServeConfigurationPlugin_1.default.accessor
            };
        }
        return this._accessor;
    }
    apply(heftSession, heftConfiguration) {
        DevDeployPlugin_1.default.apply(heftSession, heftConfiguration);
        WebpackConfigurationPlugin_1.default.apply(heftSession, heftConfiguration);
        WebpackServeConfigurationPlugin_1.default.apply(heftSession, heftConfiguration);
        CustomizeWebpackConfigurationPlugin_1.default.apply(heftSession, heftConfiguration, {
            customizationFilename: 'spfx-customize-webpack.js'
        });
    }
}
exports.default = new SpfxHeftPlugin();
