// @microsoft/sp-adaptive-card-extension-base/lib-commonjs/common/Killswitches.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isAsyncQuickViewKSActivated = exports.isIsolatedTABTestKSActivated = exports.isQuickViewLoadEngagementLogRenderTypeCardKSActivated = exports.isTargetBotDriveAceAspxKSActivated = exports.isCopyEngagementToInteranaKSActivated = exports.isGeolocationAllowedInIsolatedDomainKSActivated = exports.isLogIsInternalToGetRenderArgumentQosKSActivated = exports.isIsolatedAdaptiveCardExtensionWebPropertyPaneWorkAroundKSActivated = exports.isLogScenarioIdKSActivated = exports.isIframeReloadKSActivated = exports.isCaptureACELifecyclePerformanceKSActivated = exports.isAdditionalLoggingKSActivated = exports.isSelectMediaPropagateActionDataKSActivated = exports.isCreateMSTeamsDeepLinkForVCDKSActivated = void 0;
var sp_core_library_1 = require("@microsoft/sp-core-library");
function isCreateMSTeamsDeepLinkForVCDKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('cfe2dadf-9488-4409-b3d4-d3dd23a7970e' /* '12/08/2021', 'create msteams protocol deep link for VCD' */);
}
exports.isCreateMSTeamsDeepLinkForVCDKSActivated = isCreateMSTeamsDeepLinkForVCDKSActivated;
function isSelectMediaPropagateActionDataKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('f9047381-ce17-4042-bc46-d04681d0016b'
    /* '02/10/2022', 'Propagate correct action data to base adaptive card extension internal action handler' */
    );
}
exports.isSelectMediaPropagateActionDataKSActivated = isSelectMediaPropagateActionDataKSActivated;
function isAdditionalLoggingKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('6836bc4c-8a69-4ed9-92c0-0d6187f3acb6'
    /* '02/18/2022', 'Add additional logging for the isolated ACE service' */
    );
}
exports.isAdditionalLoggingKSActivated = isAdditionalLoggingKSActivated;
function isCaptureACELifecyclePerformanceKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('5fbd4c87-42a1-445d-8480-f1fde56ccc21'
    /* '02/18/2022', 'Add additional logging for the isolated ACE service' */
    );
}
exports.isCaptureACELifecyclePerformanceKSActivated = isCaptureACELifecyclePerformanceKSActivated;
function isIframeReloadKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('5533d556-60de-40a4-ab2e-ff72e237d89e' /* '03/09/2022', 'Reload iframe if timeout was hit once' */);
}
exports.isIframeReloadKSActivated = isIframeReloadKSActivated;
function isLogScenarioIdKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('b544c716-bfa4-4bbb-a8df-237a8ecf2e47' /* '03/09/22', 'Log scenario id as extra data' */);
}
exports.isLogScenarioIdKSActivated = isLogScenarioIdKSActivated;
function isIsolatedAdaptiveCardExtensionWebPropertyPaneWorkAroundKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('b8c50ded-d048-4eb9-bc41-21190d9d3bbb'
    /* '03/17/2022', 'Work around to enable property pane for isolated ACE web' */
    );
}
exports.isIsolatedAdaptiveCardExtensionWebPropertyPaneWorkAroundKSActivated = isIsolatedAdaptiveCardExtensionWebPropertyPaneWorkAroundKSActivated;
function isLogIsInternalToGetRenderArgumentQosKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('a95479e5-e6fb-4025-855c-a344a22d7466'
    /* '03/18/22', 'Add isInternal property to extra data for getRenderArguments QoS monitor.' */
    );
}
exports.isLogIsInternalToGetRenderArgumentQosKSActivated = isLogIsInternalToGetRenderArgumentQosKSActivated;
function isGeolocationAllowedInIsolatedDomainKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('bc68fda3-34d5-496b-8515-8d140a0c98ac'
    /* '03/21/22', 'Allow geolocation on isolated domains of ACEs.' */
    );
}
exports.isGeolocationAllowedInIsolatedDomainKSActivated = isGeolocationAllowedInIsolatedDomainKSActivated;
function isCopyEngagementToInteranaKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('8af93b94-0baf-43b2-9f46-c302729482e8'
    /* '04/11/22', 'Add argument to qos monitor to propagate data to interana/scuba.' */
    );
}
exports.isCopyEngagementToInteranaKSActivated = isCopyEngagementToInteranaKSActivated;
function isTargetBotDriveAceAspxKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('e8daf7db-132e-4225-93a4-70cc5fc06bb9'
    /* '04/14/22', 'Target botDriveAce.aspx for Bot Drive ACEs.' */
    );
}
exports.isTargetBotDriveAceAspxKSActivated = isTargetBotDriveAceAspxKSActivated;
function isQuickViewLoadEngagementLogRenderTypeCardKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('6d5c9ddb-5bc7-4b59-bbdf-ff9052d33daf'
    /* '05/04/22', 'Force engagement log renderType to Card if open QuickView action is performed.' */
    );
}
exports.isQuickViewLoadEngagementLogRenderTypeCardKSActivated = isQuickViewLoadEngagementLogRenderTypeCardKSActivated;
function isIsolatedTABTestKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('edddfc6b-79f6-49a3-8bb9-bd5a7a0eb10c'
    /* '05/10/22', 'In isolation, add data-automation-id to quick view iframe for TAB test.' */
    );
}
exports.isIsolatedTABTestKSActivated = isIsolatedTABTestKSActivated;
function isAsyncQuickViewKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('db1f6911-ac1f-42e4-890c-2c12bc4e2e00'
    /* '06/17/22', 'Allow for async loading of quickview code.' */
    );
}
exports.isAsyncQuickViewKSActivated = isAsyncQuickViewKSActivated;
